core/src/data/mbtilesDataSource.cpp

```diff
@@ -0,0 +1,422 @@
+    }
+
+    if (!m_db) { return false; }
+
+    if (_task->rawSource == this->level) {
+
+        m_worker->enqueue([this, _task, _cb](){
+            TileID tileId = _task->tileId();
+
+            auto& task = static_cast<DownloadTileTask&>(*_task);
+            task.rawTileData = std::make_shared<std::vector<char>>();
+
+            getTileData(tileId, *task.rawTileData);
+
+            if (task.hasData()) {
+                LOGW("loaded tile: %s, %d", tileId.toString().c_str(), task.rawTileData->size());
+
+                _cb.func(_task);
+
+            } else if (next) {
+
+                // Don't try this source again
+                _task->rawSource = next->level;
+
+                if (!loadNextSource(_task, _cb)) {
+                    // Trigger TileManager update so that tile will be
+                    // downloaded next time.
+                    _task->setNeedsLoading(true);
+                    requestRender();
+                }
+            } else {
+                LOGW("missing tile: %s, %d", _task->tileId().toString().c_str());
+            }
+        });
+        return true;
+    }
+
+    return loadNextSource(_task, _cb);
+}
+
+bool MBTilesDataSource::loadNextSource(std::shared_ptr<TileTask> _task, TileTaskCb _cb) {
+    if (!next) { return false; }
+
+    if (!m_db) {
+        return next->loadTileData(_task, _cb);
+    }
+
+    // Intercept TileTaskCb to store result from next source.
+    TileTaskCb cb{[this, _cb](std::shared_ptr<TileTask> _task) {
+
+        if (_task->hasData()) {
+
+            m_worker->enqueue([this, _task](){
+
+                auto& task = static_cast<DownloadTileTask&>(*_task);
+
+                LOGW("store tile: %s, %d", _task->tileId().toString().c_str(), task.hasData());
+
+                storeTileData(_task->tileId(), *task.rawTileData);
+            });
+            _cb.func(_task);
+
+        } else if (m_offlineMode) {
+            LOGW("try fallback tile: %s, %d", _task->tileId().toString().c_str());
+
+            m_worker->enqueue([this, _task, _cb](){
+
+                auto& task = static_cast<DownloadTileTask&>(*_task);
+                task.rawTileData = std::make_shared<std::vector<char>>();
+
+                getTileData(_task->tileId(), *task.rawTileData);
+
+                LOGW("loaded tile: %s, %d", _task->tileId().toString().c_str(), task.rawTileData->size());
+
+                _cb.func(_task);
+
+            });
+        } else {
+            LOGW("missing tile: %s, %d", _task->tileId().toString().c_str());
+            _cb.func(_task);
+        }
+    }};
+
+    return next->loadTileData(_task, cb);
+}
+
+void MBTilesDataSource::setupMBTiles() {
+    try {
+
+        auto mode = SQLite::OPEN_READONLY;
+
+        if (m_offlineMode) {
+            // Need to explicitly open a SQLite DB with OPEN_READWRITE
+            // and OPEN_CREATE flags to make a file and write.
+            mode = SQLite::OPEN_READWRITE | SQLite::OPEN_CREATE;
```

Comment:
yeah, there are a few modes to consider.  

1. only use existing tiles
2. try existing tiles, download and save
3. try download, fallback to existing tiles
4. try download and save to mbtiles, fallback to existing tiles

So far I intended to recreate the logic of the core-mbtiles branch. 

oh - in this case I just wanted to make sure not to overwrite mbtiles packs :)